src/shape.rs

```diff
@@ -660,237 +677,214 @@ impl ShapeLine {
+                                    fit_x -= glyph_width;
+                                    word_range_width += glyph_width;
                                     continue;
                                 } else {
-                                    word_ranges.push((
+                                    self.add_to_visual_line(
+                                        &mut current_visual_line,
+                                        span_index,
                                         fitting_start,
                                         (i, glyph_i),
                                         word_range_width,
                                         number_of_blanks,
-                                    ));
+                                    );
+                                    visual_lines.push(current_visual_line);
+                                    current_visual_line = VisualLine::default();
+
                                     number_of_blanks = 0;
-                                    fit_x = line_width - glyph_size;
-                                    word_range_width = glyph_size;
+                                    fit_x = line_width - glyph_width;
+                                    word_range_width = glyph_width;
                                     fitting_start = (i, glyph_i);
                                 }
                             }
                         } else {
                             // Wrap::Word
                             let mut prev_word_width = None;
-                            if word.blank && number_of_blanks > 0 {
+                            if word.blank {
                                 // current word causing a wrap is a space so we ignore it
-                                number_of_blanks -= 1;
-                            } else if let Some(previous_word) = span.words.get(i - 1) {
-                                // Current word causing a wrap is not whitespace, so we ignore the
-                                // previous word if it's a whitespace
-                                if previous_word.blank {
-                                    number_of_blanks -= 1;
-                                    prev_word_width = Some(previous_word.x_advance * font_size);
+                                // number_of_blanks = number_of_blanks.saturating_sub(1);
+                            } else if i > 0 {
+                                if let Some(previous_word) = span.words.get(i - 1) {
+                                    // Current word causing a wrap is not whitespace, so we ignore the
+                                    // previous word if it's a whitespace
+                                    if previous_word.blank {
+                                        prev_word_width = Some(previous_word.x_advance * font_size);
+                                        number_of_blanks = number_of_blanks.saturating_sub(1);
+                                    }
```

Comment:
Why does it matter here that the current word is blank, but not in the incongruent direction case?

I also think this should be collapsed into one a single `if !word.blank && i > 0`.

Reply:
Hopefully one day we can combine normal if and if-let together. Until then I think this is fine.

Reply:
I agree, but I was only talking about the empty `if word.blank` case. So that 
```rust
if word.blank {
} else if i > 0 {
   ...
}
```
becomes
```rust
if !word.blank && i > 0 {
    ...
}
```
I also now think the `word.blank` check is unnecessary (we don't do it in the incongruent direction case), so it should probably be just
 ```rust
if i > 0 {
    ...
}
```

Reply:
Thanks for catching this. I forgot to remove the extra `if word.blank`. I removed it in the other direction. But forgot this one.